File: lab2/python/progress_bar.py

from sys import stderr
from time import time
from typing import Iterator, Iterable, TypeVar
from types import TracebackType


T = TypeVar("T")

class ProgressBar(Iterable[T]):
    """
    A simple progress bar, inspired by the `tqdm` module.
    For "real" programs, consider installing `tqdm` instead,
    it has many more features (such as nested progress bars).
    """
    # Class-variable to turn on/off all progress bars.
    visible: bool = True

    # Instance variables.
    iterator: Iterator[T]
    description: str
    startTime: float
    total: int
    n: int
    interval: int
    unit: int
    unitSuffix: str
    barWidth: int
    descrWidth: int

    def __init__(self, iterable: Iterable[T] = (), description: str = "Logging", 
                 total: int = 0, unit: int = 0, 
                 barWidth: int = 40, descrWidth: int = 25):
        self.startTime = time()
        self.iterator = iter(iterable)
        self.description = description
        self.total = total
        if not self.total:
            self.total = len(iterable)  # type: ignore
        self.n = 0
        self.interval = max(1, min(self.total//200, 100))
        if unit == 0:
            unit = 1 if self.total < 20_000 else 1_000 if self.total < 20_000_000 else 1_000_000
        self.unit = unit
        if unit == 1:
            self.unitSuffix = " "
        elif unit == 1_000:
            self.unitSuffix = "k"
        elif unit == 1_000_000:
            self.unitSuffix = "M"
        else:
            raise ValueError("Can only handle unit == 1 or 1000 or 1_000_000")
        self.barWidth = barWidth
        self.descrWidth = descrWidth
        self._printInfoline()

    def __iter__(self) -> Iterator[T]:
        return self

    def __next__(self) -> T:
        try:
            el = next(self.iterator)
        except StopIteration:
            self._printInfoline()
            self._closeInfoline()
            raise
        self.n += 1
        if self.n % self.interval == 0:
            self._printInfoline()
        return el

    def __enter__(self) -> 'ProgressBar[T]':
        return self

    def __exit__(self, exc_type: BaseException, exc_val: BaseException, exc_tb: TracebackType):
        self._printInfoline()
        self._closeInfoline()
        pass

    def setValue(self, n: int):
        """Sets a new value for the counter. It has to be larger than the old value."""
        if n > self.n:
            self.n = n
            if n % self.interval == 0:
                self._printInfoline()

    def update(self, add: int):
        """Increases the counter by a give number."""
        self.n += add
        if self.n % self.interval == 0:
            self._printInfoline()

    def _printInfoline(self):
        if ProgressBar.visible:
            percent = 0.0
            if self.total > 0:
                percent = self.n / self.total
            hashes = round(percent * self.barWidth)
            pbar = "[" + "=" * hashes + "·" * (self.barWidth-hashes) + "]"
            elapsed = time() - self.startTime
            print(
                f"{self.description:{self.descrWidth}s} {percent:4.0%} {pbar} "
                f"{self.n/self.unit:6.0f}{self.unitSuffix} "
                f"of {self.total/self.unit:.0f}{self.unitSuffix}  | {elapsed:6.1f} s",
                file=stderr, end="\r", flush=True
            )

    def _closeInfoline(self):
        if ProgressBar.visible:
            print(file=stderr)

File: lab2/java/ProgressBar.java

import java.util.AbstractList;
import java.util.Collection;
import java.util.Iterator;


/** 
 * A simple progress bar, inspired by Python's `tqdm` module.
 * You don't have to understand how it works!
 */
class ProgressBar<T> implements Iterable<T> {

    // Class-variable to turn on/off all progress bars.
    public static boolean visible = true;

    // Instance variables.
    Iterable<T> iterable;
    String description;
    double startTime;
    int total;
    int n;
    int interval;
    int unit = 1_000;
    String unitSuffix = "k";
    int barWidth = 40;
    int descrWidth = 25;

    public static ProgressBar<Integer> range(int start, int end, String description) {
        Collection<Integer> collection = new AbstractList<>() {
            @Override
            public Integer get(int index) {
                return start + index;
            }
            @Override
            public int size() {
                return end - start;
            }
        };
        return new ProgressBar<>(collection, description);
    }

    ProgressBar(Collection<T> elements, String description) {
        this(elements, elements.size(), description);
    }

    ProgressBar(int total, String description) {
        this(null, total, description);
    }

    ProgressBar(Iterable<T> iterable, int total, String description) {
        this.startTime = System.nanoTime();
        this.iterable = iterable;
        this.description = description;
        this.total = total;
        this.n = 0;
        this.interval = Math.max(1, Math.min(this.total/200, 100));
        this.setUnit(this.total < 20_000 ? 1 : this.total < 20_000_000 ? 1_000 : 1_000_000);
        this.printInfoline();
    }

    public ProgressBar<T> setUnit(int unit) {
        this.unit = unit;
        switch (unit) {
            case 1: this.unitSuffix = " "; break;
            case 1_000: this.unitSuffix = "k"; break;
            case 1_000_000: this.unitSuffix = "M"; break;
            default: throw new IllegalArgumentException("Can only handle unit == 1 or 1000 or 1_000_000");
        }
        return this;
    }

    public ProgressBar<T> setBarWidth(int barWidth) {
        this.barWidth = barWidth;
        return this;
    }

    public ProgressBar<T> setDescrWidth(int descrWidth) {
        this.descrWidth = descrWidth;
        return this;
    }

    public Iterator<T> iterator() {
        Iterator<T> iter = iterable.iterator();
        return new Iterator<>() {
            @Override
            public boolean hasNext() {
                boolean has = iter.hasNext();
                if (!has) {
                    printInfoline();
                    closeInfoline();
                }
                return has;
            }
            @Override
            public T next() {
                n++;
                if (n % interval == 0) {
                    printInfoline();
                }
                return iter.next();
            }
        };
    }

    public void close() {
        this.printInfoline();
        this.closeInfoline();
    }

    public void setValue(int n) {
        if (n > this.n) {
            this.n = n;
            if (n % this.interval == 0) {
                this.printInfoline();
            }
        }
    }

    public void update(int add) {
        this.n += add;
        if (this.n % this.interval == 0) {
            this.printInfoline();
        }
    }

    private void printInfoline() {
        if (ProgressBar.visible) {
            float percent = 0;
            if (this.total > 0) {
                percent = (float) this.n / this.total;
            }
            int hashes = Math.round(percent * this.barWidth);
            String pbar = "[" + "=".repeat(hashes) + "·".repeat(this.barWidth - hashes) + "]";
            double elapsed = (System.nanoTime() - this.startTime) / 1e9;
            System.err.format(
                "%-" + this.descrWidth + "s %3.0f%% %s %6.0f%s of %.0f%s  | %6.1f s\r",
                this.description, 100 * percent, pbar, 
                (float) this.n/this.unit, this.unitSuffix,
                (float) this.total/this.unit, this.unitSuffix, elapsed
            );
            System.err.flush();
        }
    }

    private void closeInfoline() {
        if (ProgressBar.visible) {
            System.err.println();
        }
    }

}

